### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by EverEcheverri** [tusdotnet/tusdotnet#75]
Hi Guys, I have this behavior:
When I'm uploading a media, the memory never released, and I have to restart the service for that. in visual studio we can see that as well.

I have a simple .net core application.

>  app.UseTus(httpContext => new DefaultTusConfiguration
            {
                Store = new TusDiskStore(@"c:\tusfiles\uploads"),
                UrlPath = "/files"
            });

![image](https://user-images.githubusercontent.com/10034383/69659081-77663500-104b-11ea-916a-da9ac1fd9bd7.png)

**Comment by smatsson** [comment]
Interesting, thanks for reporting! Does the memory continue to increase over time or does it go flat at ~883 MB?

**Comment by EverEcheverri** [comment]
> Interesting, thanks for reporting! Does the memory continue to increase over time or does it go flat at ~883 MB?

In my deployed environment, after a hard night of upload media, the memory reaches 12GB, I waited for 3 hours checking if at some point the memory it is stabilized, but that not happened, still using 12GB until I had to restart the service.

![tusio](https://user-images.githubusercontent.com/10034383/69726345-599ddc00-10ee-11ea-9962-711f68a019e6.PNG)

**Comment by smatsson** [comment]
Not great... I'll do some profiling and see if I can find anything. Just to remove any other source of this issue, are you running anything else in the same server or is it just tusdotnet?

**Comment by EverEcheverri** [comment]
Only the tusdotnet, I was running a transcoding service but I turned off it and I had the same behavior.

**Comment by smatsson** [comment]
Just an update: Spent some time yesterday profiling and I think the issue is how the read and write buffers are initialized and used. I have started on a fix but won't be able to work on this anymore this week. Current status is that the memory leak seems to be gone and memory consumption is down but the files written to disk are corrupted. Glass half full I guess ;)

**Comment by smatsson** [comment]
TL; DR: A new version will be out once this issue has been fixed for real. As a workaround for now set buffer sizes to 40960 bytes (40 KB) to prevent this issue. Note that this value should be used for both the read and write buffer.
```
return new DefaultTusConfiguration
            {
                UrlPath = "/files",
                Store = new TusDiskStore(@"C:\tusfiles\", deletePartialFilesOnConcat: false, bufferSize: new TusDiskBufferSize(40960, 40960))
            };
```

Longer version:
Basically the problem is that we use a MemoryStream with an initial capacity as the write buffer. The MemoryStream is backed by an array. Once the MemoryStream's backing array gets near the size of the initial capacity the array will be resized into double its size. The default write buffer size is 50 KB meaning that the doubled size will be 100 KB. Objects larger than ~83 KB are placed by .NET in the Large Object Heap (LOH, https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/large-object-heap). These values are not GCed and compacted the same way as smaller objects, causing the memory leak. A new version is underway where we use array pooling instead allowing us to re-use arrays and not initiate new ones.

**Comment by EverEcheverri** [comment]
Thanks, I'm going to set the buffer size while the new version comes out.

**Comment by smatsson** [comment]
New version published on nuget: https://www.nuget.org/packages/tusdotnet/2.2.2

Release notes: https://github.com/tusdotnet/tusdotnet/releases/tag/2.2.2
